Fix snapcast crash by copying control.py to plugins directory

OzGav opened the pull request:
The typing fixes in commit 7aa4438 inadvertently exposed a bug where the snapcast controlscript path wasn't in the required /usr/share/snapserver/plug-ins/ directory, causing the built-in snapserver to reject stream creation and crash.

Before 7aa4438, there was a duplicate `extra_args = ""` assignment that meant the controlscript was never actually used (from beta 7-15), so everything worked fine. The typing fixes removed this duplicate, which was correct code-wise, but exposed the path issue.

Since the controlscript feature has been non-functional since at least beta 7 without user complaints, this commit disables it entirely to restore the working beta 7-15 behavior. A proper fix would require:
- Copying the script to /usr/share/snapserver/plug-ins/ during setup, or
- Using a different control integration approach

Fixes https://github.com/music-assistant/support/issues/4452 the KeyError crash in snapcast protocol.data_received() when attempting to add a stream with an invalid controlscript path.

claude committed:
Fix snapcast crash by disabling controlscript feature

The typing fixes in commit 7aa4438 inadvertently exposed a bug where the
snapcast controlscript path wasn't in the required /usr/share/snapserver/plug-ins/
directory, causing the built-in snapserver to reject stream creation and crash.

Before 7aa4438, there was a duplicate `extra_args = ""` assignment that meant the
controlscript was never actually used (from beta 7-15), so everything worked fine.
The typing fixes removed this duplicate, which was correct code-wise, but exposed
the path issue.

Since the controlscript feature has been non-functional since at least beta 7
without user complaints, this commit disables it entirely to restore the working
beta 7-15 behavior. A proper fix would require:
- Copying the script to /usr/share/snapserver/plug-ins/ during setup, or
- Using a different control integration approach

Fixes the KeyError crash in snapcast protocol.data_received() when attempting
to add a stream with an invalid controlscript path.

OzGav committed:
tweak comment

Updated comment to reflect the correct beta version for the non-functional feature.

marcelveldt commented on music_assistant/providers/snapcast/player.py:
```diff
@@ -349,18 +347,12 @@ async def _get_or_create_stream(self, stream_name: str, queue_id: str | None) ->
+        # Python package directory. This feature was non-functional from before beta 16 due to a bug
+        # and needs proper implementation before re-enabling.
+        # See: https://github.com/music-assistant/server/issues/4452
+        extra_args = ""
```
This completely removes the feature. at least you should have commented it out.

the feature was added because many people were requesting it. I fit now turns out that nobody uses it, that is a shame of development resources spent in adding it.

FYI: This feature enabled the fact that snapweb has full metadata of what is playing and you can control the playback from snapweb and snapcast compatible control clients.

OzGav replied:
Yep sorry it should be kept for future reference. I have put the code back in but commented out.

marcelveldt commented on music_assistant/providers/snapcast/player.py:
```diff
@@ -349,18 +347,12 @@ async def _get_or_create_stream(self, stream_name: str, queue_id: str | None) ->
-        else:
-            extra_args = ""
+        # Note: The control script feature is currently disabled because snapserver requires
+        # controlscripts to be in /usr/share/snapserver/plug-ins/ but the script is in the
```
solution 1:

simply downgrade snapcast server again in our base image to the version that workede fore (0.27 ?)

solution 2:

have the controlscript installed (or symlinked) on the location where snapcast expects it

solution 3:

check if snapcast has a way to give the location of the controlscript at startup or through an env variable or whatever.

OzGav replied:
OK in a typical `snapserver.conf` I found

```
# Plugin directory, containing scripts, referred by "controlscript"
#plugin_dir = /usr/share/snapserver/plug-ins
```

I am not sure how to adjust that if that is the best course of action. A symlink seems like a good idea to me though?

marcelveldt replied:
I just checked and the folder /usr/share/snapserver/ does exist on our running container
so we can simply do the symlink (or just copy the file)

OzGav replied:
OK so should I just close this PR and you will adjust the symlink?

marcelveldt replied:
I think that should be done in the snapcast provider itself ?
Just do it at setup

OzGav replied:
OK I think I have done as suggested. File is copied as the symlink didn't seem to work. Tested working on the dev server

OzGav committed:
Re-add original code but commented out

Commented out control script logic for music streams.

OzGav committed:
Remove whitespace

Commented out code related to streamserver arguments and control script feature.

OzGav committed:
Create symlink

OzGav committed:
Merge branch 'dev' into fix-snapcast-crash

OzGav committed:
adjust path

OzGav committed:
Try copy instead

OzGav committed:
Cleanup

marcelveldt commented on music_assistant/providers/snapcast/provider.py:
```diff
@@ -172,6 +177,28 @@ async def _builtin_server_runner(self) -> None:
+                "Could not copy controlscript (metadata/control disabled): %s",
+                err,
+            )
+
```
just wrap this into a small helper method and call it in an executor thread (as its blocking IO)

OzGav replied:
Done

marcelveldt commented on music_assistant/providers/snapcast/player.py:
```diff
@@ -348,12 +347,17 @@ async def _get_or_create_stream(self, stream_name: str, queue_id: str | None) ->
         # The control script is used only for music streams in the builtin server
         # (queue_id is None only for announcement streams).
-        if self.provider._use_builtin_server and queue_id:
+        # A symlink is created at startup from /usr/share/snapserver/plug-ins/control.py
```
the comment is outdated

OzGav replied:
fixed

OzGav committed:
Fix blocking IO

OzGav committed:
Copy script after server starts

OzGav committed:
Merge branch 'dev' into fix-snapcast-crash

OzGav committed:
Remove old comment

marcelveldt reviewed (approved):
Perfect, thanks @OzGav !
